### PR TITLE
Download in headless mode in tests

### PR DESCRIPTION
**Pull request by boubkerbribri**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.6.x
| Description?  | Fixing tests for headless mode by : <br/> - Enable downloading in headless mode <br/> - Deleting target so we could download in the same tab
| Type?         | refacto
| Category?     | TE
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | no
| How to test?  | - TEST_PATH=full/01_orders/* npm run specific-test -- --URL=shop_URL --DIR=download_DIR <br/>  - TEST_PATH=full/16_file/* npm run specific-test -- --URL=shop_URL --DIR=download_DIR

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/13811)

### CODE REVIEW

File: tests/E2E/package.json [hunk 1]
```diff
@@ -27,6 +27,7 @@
   },
   "homepage": "https://github.com/PrestaShop/PrestaShop#readme",
   "dependencies": {
+    "got": "^9.6.0",
```
Comment (matks) [inline tests/E2E/package.json]: @PierreRambaud 👀

File: tests/E2E/test/common.webdriverio.js [hunk 2]
```diff
@@ -179,10 +179,12 @@ module.exports = {
     if (typeof global.selenium_host !== 'undefined') {
       options.host = global.selenium_host;
     }
+    else global.selenium_host = 'localhost';
```
Comment (PierreRambaud) [inline tests/E2E/test/common.webdriverio.js]: Please :smile_cat: 

```
if (typeof global.selenium_host !== 'undefined') {
      options.host = global.selenium_host;
    } else {
      // We need to set it to localhost for "got" library
      global.selenium_host = 'localhost';
    }
```

PS: Same for the port

File: tests/E2E/package.json (hunk 1 above)
Comment (PierreRambaud) [inline tests/E2E/package.json]: why not using chai-http?

File: tests/E2E/test/clients/common_client.js [hunk 3]
```diff
@@ -120,14 +121,32 @@ class CommonClient {
...
+      return headless_client;
     } else {
       return this.client.init().windowHandleSize({width: 1280, height: 1024});
     }
   }
 
+  /**
+   * Enable download in headless mode
+   */
+  async enableDownload(){
```
Comment (matks) [inline tests/E2E/test/clients/common_client.js]: If this function is only for headless mode ... what about:
- rename it `enableDownloadForHeadlessMode()` to make it very clear
- add a check inside to make sure no one tries to use it without headless mode (see example below)

```
async enableDownload(){
    if( !global.headless) {
        throw new Error('This function should never be called without headless mode');
    }
    ...
```

File: tests/E2E/test/clients/common_client.js [hunk 4]
```diff
@@ -885,6 +904,22 @@ class CommonClient {
...
   }
+
+  /**
+   * remove attribute from selector
+   * @param selector, xpath or css selector of the element
+   * @param attributeName, attribute to remove
+   * @param isXpath, true if selector is xpath, false if css selector
+   * @param pause
+   */
+  removeAttribute(selector, attributeName, isXpath = true, pause = 0) {
```
Comment (matks) [inline tests/E2E/test/clients/common_client.js]: It seems a bit strange to perform the pause inside the `removeAttribute` 🤔 why not before calling it ?

Maybe I'm missing something here

**Comment by matks** [comment]
I see that mot of the times, using `await client.removeAttribute(OrderPage.credit_slip_document_name,'target');` fixes the script for headless mode.

I have 2 suggestions:
1) What problem does it solve and why removing this attribute fixes it ? This can be covered by some  doc or comment.
2) If this statement is reused multiple times, and it will be reused more ... let's make it a function ? 😄 Something like `removeAttributeForDocumentInHeadlessMode()` ? I'm not sure about this one, but it would make the code goal very clear

File: tests/E2E/test/clients/common_client.js (hunk 4 above)
Comment (boubkerbribri) [reply tests/E2E/test/clients/common_client.js]: no need for pause in this function

File: tests/E2E/package.json (hunk 1 above)
Comment (boubkerbribri) [reply tests/E2E/package.json]: I didn't know about chai-http.
Thanks for suggestion.

File: tests/E2E/test/common.webdriverio.js (hunk 2 above)
Comment (boubkerbribri) [reply tests/E2E/test/common.webdriverio.js]: Thanks for suggestion
done

File: tests/E2E/package.json (hunk 1 above)
Comment (boubkerbribri) [reply tests/E2E/package.json]: deleted :smile:

File: tests/E2E/test/clients/common_client.js (hunk 3 above)
Comment (boubkerbribri) [reply tests/E2E/test/clients/common_client.js]: >     * rename it `enableDownloadForHeadlessMode()` to make it very clear
=> done 

>     * add a check inside to make sure no one tries to use it without headless mode (see example below)
> ```
> async enableDownload(){
>     if( !global.headless) {
>         throw new Error('This function should never be called without headless mode');
>     }
>     ...
> ```
=> No need to throw error, calling the function in HeadFull mode will not affect the test.

File: tests/E2E/test/clients/common_client.js (hunk 4 above)
Comment (boubkerbribri) [reply tests/E2E/test/clients/common_client.js]: deleted

**Comment by boubkerbribri** [comment]
> I see that mot of the times, using `await client.removeAttribute(OrderPage.credit_slip_document_name,'target');` fixes the script for headless mode.
> 
> I have 2 suggestions:
> 
>     1. What problem does it solve and why removing this attribute fixes it ? This can be covered by some  doc or comment.
=> Where exactly can I add doc or comments for this function ?

>     2. If this statement is reused multiple times, and it will be reused more ... let's make it a function ? smile Something like `removeAttributeForDocumentInHeadlessMode()` ? I'm not sure about this one, but it would make the code goal very clear
=> The function removeAttributeForDocumentInHeadlessMode could be used in other test (in the future), that's why I prefer to keep the name without `ForDocumentInHeadlessMode`

**Comment by matks** [comment]
> => Where exactly can I add doc or comments for this function ?

One solution would be next to the places where it's used
```
// for headless, we need to remove attribute because [...]
if(global.headless)  await client.removeAttribute(OrderPage.view_delivery_slip,'target');
```

Apart from that, code is 👍 🎉

**Review by matks — APPROVED** [review]

**Review by Quetzacoalt91 — APPROVED** [review]
I'm approving although there a few lines which could be changed as they're not following the coding standards. For the migration, it would be great to add a linter.

**Pull request merged**